x/leverage/keeper/msg_server_test.go

```diff
@@ -172,3 +173,1545 @@ func (s *IntegrationTestSuite) TestUpdateRegistry() {
+			require.Equal(tc.expectedTokens, resp.Received, tc.msg)
+
+			// final state
+			fBalance := app.BankKeeper.GetAllBalances(ctx, tc.addr)
+			fCollateral := app.LeverageKeeper.GetBorrowerCollateral(ctx, tc.addr)
+			fUTokenSupply := app.LeverageKeeper.GetAllUTokenSupply(ctx)
+			fExchangeRate := app.LeverageKeeper.DeriveExchangeRate(ctx, denom)
+			fBorrowed := app.LeverageKeeper.GetBorrowerBorrows(ctx, tc.addr)
+
+			// verify token balance increased by the expected amount
+			require.Equal(iBalance.Add(tc.expectedTokens).Sub(tc.expectFromBalance...),
+				fBalance, tc.msg, "token balance")
+			// verify uToken collateral decreased by the expected amount
+			s.requireEqualCoins(iCollateral.Sub(tc.expectFromCollateral...), fCollateral, tc.msg, "uToken collateral")
+			// verify uToken supply decreased by the expected amount
+			require.Equal(iUTokenSupply.Sub(tc.uToken), fUTokenSupply, tc.msg, "uToken supply")
+			// verify uToken exchange rate is unchanged
+			require.Equal(iExchangeRate, fExchangeRate, tc.msg, "uToken exchange rate")
+			// verify borrowed coins are unchanged
+			require.Equal(iBorrowed, fBorrowed, tc.msg, "borrowed coins")
+
+			// check all available invariants
+			s.checkInvariants(tc.msg)
+		}
+	}
+}
+
+func (s *IntegrationTestSuite) TestMsgMaxWithdraw() {
+	type testCase struct {
+		msg                  string
+		addr                 sdk.AccAddress
+		denom                string
+		expectedWithdraw     sdk.Coin
+		expectFromCollateral sdk.Coin
+		expectedTokens       sdk.Coin
+		err                  error
+	}
+
+	app, ctx, srv, require := s.app, s.ctx, s.msgSrvr, s.Require()
+
+	// create and fund a supplier with 100 UMEE and 100 ATOM, then supply 100 UMEE and 50 ATOM
+	// also collateralize 75 of supplied UMEE
+	supplier := s.newAccount(coin(umeeDenom, 100_000000), coin(atomDenom, 100_000000))
+	s.supply(supplier, coin(umeeDenom, 100_000000))
+	s.collateralize(supplier, coin("u/"+umeeDenom, 75_000000))
+	s.supply(supplier, coin(atomDenom, 50_000000))
+
+	// create and modify a borrower to force the uToken exchange rate of ATOM from 1 to 1.2
+	borrower := s.newAccount(coin(atomDenom, 100_000000))
+	s.supply(borrower, coin(atomDenom, 100_000000))
+	s.collateralize(borrower, coin("u/"+atomDenom, 100_000000))
+	s.borrow(borrower, coin(atomDenom, 10_000000))
+	s.tk.SetBorrow(ctx, borrower, coin(atomDenom, 40_000000))
+
+	// create an additional UMEE supplier with a small borrow
+	other := s.newAccount(coin(umeeDenom, 100_000000))
+	s.supply(other, coin(umeeDenom, 100_000000))
+	s.collateralize(other, coin("u/"+umeeDenom, 100_000000))
+	s.borrow(other, coin(umeeDenom, 10_000000))
+
+	// create an additional supplier (UMEE, DUMP, PUMP tokens)
+	surplus := s.newAccount(coin(umeeDenom, 100_000000), coin(dumpDenom, 100_000000), coin(pumpDenom, 100_000000))
+	s.supply(surplus, coin(umeeDenom, 100_000000))
+	s.supply(surplus, coin(pumpDenom, 100_000000))
+	s.supply(surplus, coin(dumpDenom, 100_000000))
```

Comment:
There are differences in some of the setups - since there is bound to be repetition in unit test setup, I might not be able to reduce all of it.

I'll think about giving `s.supply` and others the ability to accept `sdk.Coin...` args to eliminate some of this.